### PR TITLE
allow defensive operations to take a proof

### PR DESCRIPTION
**Pull request by kianenigma**
I didn't do it quite delicately, but gets the job done for now. 

Open to an more flexible alternative, specially keeping `defensive!` flexible. 

take out of https://github.com/paritytech/substrate/pull/11013/files.

I changed one or two examples to demonstrate how it should be used.

### CODE REVIEW

File: frame/support/src/traits/misc.rs [hunk 1]
```diff
@@ -102,6 +112,10 @@ pub trait Defensive<T> {
 	/// }
 	/// ```
 	fn defensive(self) -> Self;
+
+	/// Same as [`Defensive::defensive`], but it takes a proof is input, and displays it if the
```
Comment (ggwpez) [inline frame/support/src/traits/misc.rs:116]:
```suggestion
	/// Same as [`Defensive::defensive`], but it takes a proof as input, and displays it if the
```
```suggestion
	/// Same as [`Defensive::defensive`], but it takes a proof is input, and displays it if the
```

File: frame/bags-list/src/list/mod.rs [hunk 2]
```diff
@@ -443,8 +443,7 @@ impl<T: Config<I>, I: 'static> List<T, I> {
 
 		// remove the heavier node from this list. Note that this removes the node from storage and
 		// decrements the node counter.
-		// defensive: both nodes have been checked to exist.
-		let _ = Self::remove(&heavier_id).defensive();
+		let _ = Self::remove(&heavier_id).defensive("both nodes have been checked to exist; qed");
```
Comment (ggwpez) [inline frame/bags-list/src/list/mod.rs:446]: Alternatively could we do?  
```suggestion
		let _ = Self::remove(&heavier_id).with_proof("both nodes have been checked to exist; qed").defensive();
```
```suggestion
		let _ = Self::remove(&heavier_id).defensive("both nodes have been checked to exist; qed");
```

Comment (ggwpez) [inline frame/bags-list/src/list/mod.rs:446]: Alternatively could we do?  
```suggestion
		let _ = Self::remove(&heavier_id).with_proof("both nodes have been checked to exist; qed").defensive();
```
```suggestion
		let _ = Self::remove(&heavier_id).defensive("both nodes have been checked to exist; qed");
```

File: frame/staking/src/slashing.rs [hunk 3]
```diff
@@ -600,8 +600,8 @@ pub fn do_slash<T: Config>(
 	slashed_imbalance: &mut NegativeImbalanceOf<T>,
 	slash_era: EraIndex,
 ) {
-	let controller = match <Pallet<T>>::bonded(stash) {
-		None => return, // defensive: should always exist.
+	let controller = match <Pallet<T>>::bonded(stash).defensive() {
```
Comment (ggwpez) [inline frame/staking/src/slashing.rs:603]: Why did you not add the `should always exist` here as reason?

**Review by ggwpez — APPROVED** [review]

File: frame/bags-list/src/list/mod.rs (hunk 2 above)
Comment (ggwpez) [reply frame/bags-list/src/list/mod.rs:446]: sorry slow internet

Comment (kianenigma) [reply frame/bags-list/src/list/mod.rs:446]: yeah, but would require another extension trait on Result and Option, justified? not sure.

File: frame/staking/src/slashing.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (kianenigma) [reply frame/staking/src/slashing.rs:603]: honestly, because it is not a good proof :D

**Review by shawntabrizi — APPROVED** [review]

File: frame/bags-list/src/list/mod.rs (hunk 2 above)
Comment (ggwpez) [reply frame/bags-list/src/list/mod.rs:446]: Probably overkill for now.

Comment (KiChjang) [inline frame/bags-list/src/list/mod.rs:446]:
```suggestion
		let _ = Self::remove(&heavier_id).defensive_proof("both nodes have been checked to exist; qed");
```

File: frame/support/src/traits/misc.rs [hunk 4]
```diff
@@ -184,6 +198,16 @@ impl<T> Defensive<T> for Option<T> {
...
 	}
+
+	fn defensive_proof(self, proof: &'static str) -> Self {
+		match self {
+			Some(inner) => Some(inner),
+			None => {
+				defensive!(proof);
+				None
+			},
+		}
```
Comment (KiChjang) [inline frame/support/src/traits/misc.rs:209]:
```suggestion
		if self.is_none() {
			defensive!(proof)
		}
		self
```

File: frame/support/src/traits/misc.rs [hunk 5]
```diff
@@ -229,6 +253,16 @@ impl<T, E: sp_std::fmt::Debug> Defensive<T> for Result<T, E> {
...
 	}
+
+	fn defensive_proof(self, proof: &'static str) -> Self {
+		match self {
+			Ok(inner) => Ok(inner),
+			Err(e) => {
+				defensive!(e, proof);
+				Err(e)
+			},
+		}
```
Comment (KiChjang) [inline frame/support/src/traits/misc.rs:264]:
```suggestion
		if self.is_err() {
			defensive!(e, proof);
		}
		self
```

**Review by KiChjang — APPROVED** [review]
Couple of nits and a compilation error, then we can merge this.

**Comment by kianenigma** [comment]
bot merge

**Comment by xlc** [comment]
One minor issue of those proof string is that I believe they are included into the wasm binary, which just bloat the wasm size further. I didn't check if that's the actual case so maybe it is already optimized out.

**Comment by kianenigma** [comment]
hmm yeah that's a good point, they should not be used if that's the case then. The proof can be a comment in the code.

**Pull request merged**